src/report.rs

```diff
@@ -0,0 +1,340 @@
+use crate::context::Context;
+use csv::Writer;
+use std::any::TypeId;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::fs::File;
+use std::path::Path;
+use std::path::PathBuf;
+
+pub struct ConfigReportOptions {
+    pub file_prefix: String,
+    pub directory: String,
+}
+
+impl ConfigReportOptions {
+    #[must_use]
+    pub fn new() -> Self {
+        ConfigReportOptions {
+            file_prefix: String::new(),
+            directory: "./".to_string(),
+        }
+    }
+}
+
+impl Default for ConfigReportOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub trait Report: 'static {
+    fn type_id(&self) -> TypeId;
+    fn serialize(&self, writer: &mut Writer<File>);
+}
+
+#[macro_export]
+macro_rules! create_report_trait {
+    ($name:ident) => {
+        impl Report for $name {
+            fn type_id(&self) -> TypeId {
+                TypeId::of::<$name>()
+            }
+
+            fn serialize(&self, writer: &mut Writer<File>) {
+                writer.serialize(self).unwrap();
+            }
+        }
+    };
+}
+
+struct ReportData {
+    file_writers: RefCell<HashMap<TypeId, Writer<File>>>,
+    config: ConfigReportOptions,
+}
+
+crate::context::define_data_plugin!(
+    ReportPlugin,
+    ReportData,
+    ReportData {
+        file_writers: RefCell::new(HashMap::new()),
+        config: ConfigReportOptions::new(),
+    }
+);
+
+pub trait ContextReportExt {
+    fn generate_filename(&mut self, short_name: &str) -> PathBuf;
+    fn add_report<T: Report + 'static>(&mut self, short_name: &str);
+    fn send_report<T: Report>(&mut self, report: T);
+    fn init_reports(&mut self, file_prefix: Option<&str>, directory: Option<&str>);
+}
+
+impl ContextReportExt for Context {
+    fn generate_filename(&mut self, short_name: &str) -> PathBuf {
```

Comment:
Rather than making this part of the trait extension, can you make this a private helper function?